### ISSUE TITLE
Callout with image: Image not shawn on Android

### ISSUE THREAD

**Issue by slorber** [react-native-maps/react-native-maps#2633]
### Is this a bug report?

Yes

### Have you read the [Installation Instructions](https://github.com/react-community/react-native-maps/blob/master/docs/installation.md)?

Yes

### Environment

Expo SDK 31 (includes maps v0.22.1)

Target device: Samsung Galaxy S9 hardware

### Steps to Reproduce

Use a marker callout with an image

### Expected Behavior

The image should render

### Actual Behavior

The image does not render on Android, but it renders fine on iOS

### Reproducible Demo

```js
              <Marker
                coordinate={location}
              >
                <Callout>
                  <View
                    style={{
                      width: 100,
                      height: 100,
                      backgroundColor: 'blue'
                    }}
                  >
                    <Image
                      source={{
                        uri:
                          'https://cdn.pixabay.com/photo/2016/06/18/17/42/image-1465348_1280.jpg'
                      }}
                      style={{
                        width: 50,
                        height: 50,
                        backgroundColor: 'red',
                        resizeMode: 'cover'
                      }}
                    />
                    <Text>hey</Text>
                  </View>
                </Callout>
              </Marker>
```

This will render:

![image](https://user-images.githubusercontent.com/749374/50156811-01e8d500-02d0-11e9-92b1-747ba4360939.png)

Instead of:

![image](https://user-images.githubusercontent.com/749374/50156960-6b68e380-02d0-11e9-8034-5d8e5ce3a6f9.png)

Note that:
- even the red background of the image is not shawn on Android
- it does not work either with base64 image or a required local image
- it does not work with/without tooltip
- it does not work with/without custom marker
- it does not work if I toggle the marker/callout again

**Comment by Zazzzles** [comment]
Also having this issue with exact same setup

**Comment by ilkerdogan** [comment]
I am having the same problem. When I use a static image in Callout, it works fine. But when I use an image with a base64 source in Callout, It loads too late.

**Comment by akshaygpta007** [comment]
https://github.com/react-native-community/react-native-maps/issues/1870#issuecomment-485329883


check above comment.

**Comment by kratuwu** [comment]
Is there any other way isn't Webview?

**Comment by malFuckingFurion** [comment]
Hello,
Use Image inside Text component
`<Text>
 <Image style={{ height: 100, width:100 }} source={{ ... }} resizeMode="cover" /> 
</Text>`

**Comment by aacassandra** [comment]
@malFuckingFurion strange but True 👍🏻

**Comment by dragma** [comment]
Until this is merged, you can all `yarn add https://github.com/dragma/react-native-maps.git#BOUYA` instead of the official one :)

**Comment by tzs007** [comment]
It's not working with expo 38, and react native maps 0.27.1

Any solution?

**Comment by JonatanOrtiz** [comment]
<View >
  <WebView source={{ uri: ... }} style={ height: 100, width:100 } />
</View>

**Comment by JonatanOrtiz** [comment]
...
<View>
      <WebView style={{ height: 100 , width: 230, }} source={{ uri: ... }} />
</View>

**Comment by JonatanOrtiz** [comment]
`
<View>
      <WebView style={{ height: 100 , width: 230, }} source={{ uri: ... }} />
</View>
`

**Comment by ajenkuhnabat** [comment]
> `<View> <WebView style={{ height: 100 , width: 230, }} source={{ uri: ... }} /> </View>`

works perfectly for me. thank you !!!

**Comment by RavisaraDev** [comment]
I had the same issue even i wrap the <Image/> inside `<Text/>` it showed only half of image.  but make` <Text/>` component relative and adding some bottom value fixed the issue.

```
<Text style={{ height: 200, position: "relative", bottom: 40 }}>
    <Image
      resizeMode="cover"
      style={{ width: 70, height: 90, }}
      source={{ uri: imageUrl }}
    />
</Text>
```

**Comment by MustaphaGhlissi** [comment]
Actually, rendered **Image** inside a **Text** doesn't work properly,
Same for **WebView** use case, it's not possible to control **Image** display such as **ResizeMode** inside it while **WebView** source is the image link !

**_Solution:_**
An alternative solution is to use [react-native-svg Image](https://github.com/react-native-svg/react-native-svg#image)

Code example:

          ```
  Import {Svg, Image as ImageSvg} from 'react-native-svg';

             <Svg width={240} height={120}>
                     <ImageSvg
                         width={'100%'} 
                         height={'100%'}
                         preserveAspectRatio="xMidYMid slice"
                         href={{ uri: 'http://lorempixel.com/400/200/'}}
                     />
                 </Svg>

```

**Comment by zholmes1** [comment]
> Actually, rendered **Image** inside a **Text** doesn't work properly,
> Same for **WebView** use case, it's not possible to control **Image** display such as **ResizeMode** inside it while **WebView** source is the image link !
> 
> **_Solution:_**
> An alternative solution is to use [react-native-svg Image](https://github.com/react-native-svg/react-native-svg#image)
> 
> Code example:
> 
> ```
>       Import {Svg, Image as ImageSvg} from 'react-native-svg';
> 
>          <Svg width={240} height={120}>
>                  <ImageSvg
>                      width={'100%'} 
>                      height={'100%'}
>                      preserveAspectRatio="xMidYMid slice"
>                      href={{ uri: 'http://lorempixel.com/400/200/'}}
>                  />
>              </Svg>
> ```

Great solution - this is by far the best of what's been proposed so far.

**Comment by pperron** [comment]
For my part, I followed the proposed solution from @karpyshynroman when I'm using web source images.
 
I tried with the `<Text><Image /></Text>` solution and it worked for the image that I was loading from the app resources (`source={require('...whatever.png')}`) but not for images I was fetching from the web using HTTP URI.

Using the @karpyshynroman that I tweaked a bit (pre-loaded zero width/height images before <MapView>) solved the issue for me. Here a sample of what I did:
```
 {this.state.markers.map((marker, index) => (
   <Image
     style={{
        width: 0,
        height: 0,
     }}
     source={marker.image}
     onError={this.onIMGError.bind(this)}
     resizeMode="cover"
   />
 ))}
  <MapView>
   {this.state.markers.map((marker, index) => (
    <MapView.Marker>
        <MapView.Callout>
            <Text>
                   <Image
                         resizeMode="cover"
                         style={{
                              width: 100,
                              height: 100,
                         }}
                        source={marker.image}
                     />
              </Text>
        <MapView.Callout />
 </MapView>
```
So basically, just loading the images before the <MapView> the use those same images inside the <MapView.Callout> surrounded by `<Text><Image /></Text>` seems to solve the refresh issue and loads images when callout open.

Hope this will help someone else :)

**Comment by rafaelmonroy** [comment]
> > Actually, rendered **Image** inside a **Text** doesn't work properly,
> > Same for **WebView** use case, it's not possible to control **Image** display such as **ResizeMode** inside it while **WebView** source is the image link !
> > **_Solution:_**
> > An alternative solution is to use [react-native-svg Image](https://github.com/react-native-svg/react-native-svg#image)
> > Code example:
> > ```
> >       Import {Svg, Image as ImageSvg} from 'react-native-svg';
> > 
> >          <Svg width={240} height={120}>
> >                  <ImageSvg
> >                      width={'100%'} 
> >                      height={'100%'}
> >                      preserveAspectRatio="xMidYMid slice"
> >                      href={{ uri: 'http://lorempixel.com/400/200/'}}
> >                  />
> >              </Svg>
> > ```
> 
> Great solution - this is by far the best of what's been proposed so far.

still not as smooth as iOS but at least it works!

**Comment by vassalloandrea** [comment]
Both the solutions work but I have a problem on the first tap, the callout doesn't appear. Did anyone solve this?

**Comment by rafaelmonroy** [comment]
@vassalloandrea yeah i noticed it only works on second tap. Haven’t found a solution yet.

**Comment by helloo0-0oword** [comment]
> @vassalloandrea, @rafaelmonroy
> I had the same problem...
> 
> ```
> import React, { useState, useEffect } from "react";
> import { View, } from "react-native";
> import MapView, { Callout, Marker } from "react-native-maps";
> import { Svg, Image as ImageSvg } from "react-native-svg";
> 
> const markers = [...];
> const numberOfMarkers = markers.length; // length of markers array
> const [markerRefs, setMarkerRefs] = useState([]);
> 
>   useEffect(() => {
>     setMarkerRefs((markerRefs) =>
>       Array(numberOfMarkers )
>         .fill()
>         .map((_, i) => markerRefs[i] || React.createRef())
>     );
>   }, [numberOfMarkers ]);
>   
>   ....
>   
>   return (
>     <MapView ...>
>       {markers.map((marker, index) => {
>           return (
>           <Marker
>               key={index}
>               coordinate={marker.coordinate}
>               ref={markerRefs[index]}
>               onPress={() => {
>                 setTimeout(() => {
>                   markerRefs[index].current.hideCallout();
>                   markerRefs[index].current.showCallout();
>                 }, 200);
>               }}
>             >
>               <Callout tooltip>
>                 <View style={styles.thumbnailContainer}>
>                     <Svg>
>                       <ImageSvg
>                         width={"100%"}
>                         height={"100%"}
>                         preserveAspectRatio="xMidYMid slice"
>                         href={{ uri: 'http...' }}
>                       />
>                     </Svg>
>                   </View>
>               </Callout>
>             </Marker>
>           );
>         })}
>     </MapView>
>   )
> ```
> 
> So what I'm doing here is:
> I created refs for every markers. And after a marker tap (Marker component's onPress event) it simulates another tap...
> Yes, it is a hack, but it works :D

Thks it work :D

**Comment by musman97** [comment]
> Actually, rendered **Image** inside a **Text** doesn't work properly,
> Same for **WebView** use case, it's not possible to control **Image** display such as **ResizeMode** inside it while **WebView** source is the image link !
> 
> **_Solution:_**
> An alternative solution is to use [react-native-svg Image](https://github.com/react-native-svg/react-native-svg#image)
> 
> Code example:
> 
> ```
>       Import {Svg, Image as ImageSvg} from 'react-native-svg';
> 
>          <Svg width={240} height={120}>
>                  <ImageSvg
>                      width={'100%'} 
>                      height={'100%'}
>                      preserveAspectRatio="xMidYMid slice"
>                      href={{ uri: 'http://lorempixel.com/400/200/'}}
>                  />
>              </Svg>
> ```

It is the best solution. Worked great.

**Comment by VMKLL** [comment]
> Actually, rendered **Image** inside a **Text** doesn't work properly,
> Same for **WebView** use case, it's not possible to control **Image** display such as **ResizeMode** inside it while **WebView** source is the image link !
> 
> **_Solution:_**
> An alternative solution is to use [react-native-svg Image](https://github.com/react-native-svg/react-native-svg#image)
> 
> Code example:
> 
> ```
>       Import {Svg, Image as ImageSvg} from 'react-native-svg';
> 
>          <Svg width={240} height={120}>
>                  <ImageSvg
>                      width={'100%'} 
>                      height={'100%'}
>                      preserveAspectRatio="xMidYMid slice"
>                      href={{ uri: 'http://lorempixel.com/400/200/'}}
>                  />
>              </Svg>
> ```

Works great! Thanks 🙏
Anyone know how can I round corners like if I used "BorderRadius" style for "Image"?

**Comment by Albertyhu** [comment]
Unfortunately, I tried all the ideas suggested here and they didn't work. I'm currently using expo on Android Studio to create my app.

**Comment by itsravenous** [comment]
> > Actually, rendered **Image** inside a **Text** doesn't work properly,
> > Same for **WebView** use case, it's not possible to control **Image** display such as **ResizeMode** inside it while **WebView** source is the image link !
> > **_Solution:_**
> > An alternative solution is to use [react-native-svg Image](https://github.com/react-native-svg/react-native-svg#image)
> > Code example:
> > ```
> >       Import {Svg, Image as ImageSvg} from 'react-native-svg';
> > 
> >          <Svg width={240} height={120}>
> >                  <ImageSvg
> >                      width={'100%'} 
> >                      height={'100%'}
> >                      preserveAspectRatio="xMidYMid slice"
> >                      href={{ uri: 'http://lorempixel.com/400/200/'}}
> >                  />
> >              </Svg>
> > ```
> 
> Works great! Thanks 🙏
> Anyone know how can I round corners like if I used "borderRadius" style for "Image"?

@VMKLL Yep! I needed exactly the same thing. The snippet you need actually happens to be in the linked docs: https://github.com/react-native-svg/react-native-svg#image. Included it here (with just the bits you need) in case of linkrot:

```js
<Svg height="100" width="100">
  <Defs>
    <ClipPath id="clip">
      <Circle cx="50%" cy="50%" r="40%" />
    </ClipPath>
  </Defs>

  <Image
    x="5%"
    y="5%"
    width="50%"
    height="90%"
    preserveAspectRatio="xMidYMid slice"
    opacity="0.5"
    href={require('../image.jpg')}
    clipPath="url(#clip)"
  />
```

How it works: In the "defs" component (short for definitions - a place in an SVG document where you can define stuff you want to use throughout, e.g. pattern fills, images, or in this case, a clip path) you create a circle. Then you reference that circle as the clip path for your image.

**Comment by stuartrapop** [comment]
Does anyone have a working example of a marker Callout which contains an image and works on both android and ios?  I have tried the solutions above, but have not been able to get any of them to work. 

I currently have to click twice to see the image on android.  


I am displaying thousands of points on the map with external uri's for the images.  These are the packages I am using in my test (obviously many of them are not relevant).  

Below is a test popup component.

```
"dependencies": {
    "@expo/vector-icons": "^12.0.0",
    "@mapbox/geo-viewport": "^0.4.1",
    "expo": "~42.0.1",
    "expo-status-bar": "~1.0.4",
    "react": "16.13.1",
    "react-dom": "16.13.1",
    "react-native": "https://github.com/expo/react-native/archive/sdk-42.0.0.tar.gz",
    "react-native-maps": "0.28.0",
    "react-native-svg": "12.1.1",
    "react-native-web": "~0.13.12",
    "react-native-webview": "11.6.2",
    "supercluster": "^7.1.3"
  },
```
```
export default function Popup(): ReactElement {
  return (
    <View style={styles.container}>
      <Text>Eau minérale Mélodie</Text>
      <Text style={{ width: '100%', height: '100%' }}>
        <Image
          resizeMode="cover"
          style={styles.image}
          source={{
            uri: 'https://upload.wikimedia.org/wikipedia/commons/thumb/0/05/Eau_min%C3%A9rale_M%C3%A9lodie.jpg/250px-Eau_min%C3%A9rale_M%C3%A9lodie.jpg',
          }}
        />
      </Text>
    </View>
  )
}
```

**Comment by SArsalanK** [comment]
A way around to this is using a `WebView` instead of an `Image`

```
const isAndroid = Platform.OS === 'android'
{isAndroid ? <WebView style={styles.compactImage} source={{ uri: restaurant.photos[0] }} /> : <Image style={styles.compactImage} source={{ uri: restaurant.photos[0] }} />}
```

**Comment by benohiit** [comment]
No any of the previous answers worked for me.
I guess if we simulated the tap on the marker it could work like @akos96kov said , but I dont understand how to write it:

`props.list.map(item => (
<Marker
                coordinate={{
                    latitude: parseFloat(item.latitude),
                    longitude: parseFloat(item.longitude)
                }}
onPress={() => {
                    setTimeout(() => {
                        hideCallout();                      // how could I write it ?
                        showCallout();
                    }, 200);
                }}
            >
                <CalloutDescriptions infos={item} />

            </Marker>))`

**Comment by tkyr-hh** [comment]
> Actually, rendered **Image** inside a **Text** doesn't work properly, Same for **WebView** use case, it's not possible to control **Image** display such as **ResizeMode** inside it while **WebView** source is the image link !
> 
> **_Solution:_** An alternative solution is to use [react-native-svg Image](https://github.com/react-native-svg/react-native-svg#image)
> 
> Code example:
> 
> ```
>       Import {Svg, Image as ImageSvg} from 'react-native-svg';
> 
>          <Svg width={240} height={120}>
>                  <ImageSvg
>                      width={'100%'} 
>                      height={'100%'}
>                      preserveAspectRatio="xMidYMid slice"
>                      href={{ uri: 'http://lorempixel.com/400/200/'}}
>                  />
>              </Svg>
> ```

This is the best solution so far, thanks a lot for it.

**Comment by benohiit** [comment]
> > Actually, rendered **Image** inside a **Text** doesn't work properly, Same for **WebView** use case, it's not possible to control **Image** display such as **ResizeMode** inside it while **WebView** source is the image link !
> > **_Solution:_** An alternative solution is to use [react-native-svg Image](https://github.com/react-native-svg/react-native-svg#image)
> > Code example:
> > ```
> >       Import {Svg, Image as ImageSvg} from 'react-native-svg';
> > 
> >          <Svg width={240} height={120}>
> >                  <ImageSvg
> >                      width={'100%'} 
> >                      height={'100%'}
> >                      preserveAspectRatio="xMidYMid slice"
> >                      href={{ uri: 'http://lorempixel.com/400/200/'}}
> >                  />
> >              </Svg>
> > ```
> 
> This is the best solution so far, thanks a lot for it.

and for a local image ?

**Comment by esinanturan** [comment]
This issue still exists in the latest version even. Any updates on that ? I wonder how come it has not been solved for this long.

**Comment by Lu1815** [comment]
> Actually, rendered **Image** inside a **Text** doesn't work properly, Same for **WebView** use case, it's not possible to control **Image** display such as **ResizeMode** inside it while **WebView** source is the image link !
> 
> **_Solution:_** An alternative solution is to use [react-native-svg Image](https://github.com/react-native-svg/react-native-svg#image)
> 
> Code example:
> 
> ```
>       Import {Svg, Image as ImageSvg} from 'react-native-svg';
> 
>          <Svg width={240} height={120}>
>                  <ImageSvg
>                      width={'100%'} 
>                      height={'100%'}
>                      preserveAspectRatio="xMidYMid slice"
>                      href={{ uri: 'http://lorempixel.com/400/200/'}}
>                  />
>              </Svg>
> ```

@mustapha-ghlissi Best solution found on internet so far! Its 2024 and the issue persists! I think I am going to made a PR in the official package!

**Comment by salah-ghanim** [comment]
I'm reopening this because I see a chance to fix it on the native side but it needs to be tried 

what I see is that:
 1. android's native adapter tries to load the view only once, following changes won't be tracked.
 2. image components have an onLoad function.
 3. the InfowWindowAdapter can be implemented in a way where it returns null as long as Callout is not set to be ready .
 this might work

**Comment by salah-ghanim** [comment]
Ok, I've spent couple of hours trying multiple solutions unfortunately without a success.. 

the problem is that android GoogleMaps SDK doesn't respect the concept of Views when it comes to the InfoWindows / Callout, it simply converts the View to a bitmap by drawing it on a canvas, so any dynamic content such as Image won't be rendered correctly.

https://developers.google.com/maps/documentation/android-sdk/infowindows#custom_info_windows

a solution would be to use offline images that should load synchronously, or use an overlay view to mimic the callout ie a view on top of the map view that matches the position of the marker (I haven't explored the second idea)

closing for now, anyone with a possible solution feel free to create a pull request.

for now this is a native issue with GoogleMaps Android and we can't do much about it.
